docs/Boxes.md

```diff
@@ -0,0 +1,143 @@
+
+### Combining different FC-types in expressions
+
+The example given for `Option` above is legal Dafny code because the output of the caller method `m` has the same type
+(`Option<int>`) as the output of the callee function `Find`. But what if it does not? You may be writing code that generally uses, say, `Result`,
+but happens to be using code supplied by someone else that returns an `Outcome`.
+
+When values of these types are used in expressions, the library types offer means to convert among them: each type has instance functions that
+convert to values of the other types. For example, we can rewrite the example above this way:
+
+```dafny
+import opened Dafny.Boxes
+
+function Find<T(==)>(s: seq<T>, k: int, value: T): (r: Option<int>)
+  requires 0 <= k <= |s|
+  ensures  r.Some? ==> k <= r.Extract() < |s|
+  ensures  r.Some? ==> s[r.Extract()] == value;
+  ensures  r.None? ==> forall i | k <= i < |s| :: s[i] != value
+  decreases |s| - k
+{
+  if k >= |s| then None else if s[k] == value then Some(k) else Find(s, k+1, value)
+}
+
+method m(s: seq<int>) returns (r: Result<int,string>) {
+  var value: int := *;
+  // do various calculations
+  var index: int :- Find(s, 0, value).ToResult("not found");
+  // more calculations
+  return Success(value);
+}
+```
+
+Here we used the `ToResult` function of `Option` to convert the result value to a `Result`. There are similar `ToOutcome` and `ToOption` functions in each type.
+
+But what if we need to convert to some custom FC-type? Each datatype comes with a `Map` function, which can be given as an argument a function that converts from
+one FC-type to another. For example, we could rewrite the above as
+
+```dafny
+  var index: int :- Find(s, 0, value).
+     Map((o: Option<int>) => match o case Some(v) => Success(v) 
+                                     case None => Failure("not found") );
+```
+
+One can also just use usual function application syntax. `f(x)` to convert one FC-type
+to another.
+
+### Combining different FC-types in methods
+
+The conversion functions used in the last section work syntactically because we had boxed values that were returned by expressions (function calls), to which the conversion functions could
+be applied. When a FC-value is returned by a method there is no place to call such a conversion function: the return value of the method must be captured by either `:=` or `:-`.
+
+For this situation there is one more piece of syntax to make working with boxed values convenient. Suppose `Find` in our example is a method instead of a function. Then we have
+
+```dafny
+method m(s: seq<int>) returns (r: Result<int,string>) {
+  var value: int := *;
+  // do various calculations
+  var f = (o: Option<int>) => match o case Some(v) => Success(v) 
+                                      case None => Failure("not found")
+  var index: int :- {:convert f} Find(s, 0, value);
```

Comment:
We don't currently have https://github.com/dafny-lang/dafny/issues/2436 planned for Dafny 4.0 so let's not mention any solution here.

Reply:
I'll leave it as tentatively planned.

Reply:
Revised as suggested